File: Creational/SimpleFactory/java/AnimalFactory.java
package com.designpatterns.simplefactory;

/**
 * 動物工廠
 * 簡單工廠模式：根據類型創建不同動物
 */
public class AnimalFactory {
    
    /**
     * 創建動物
     * @param animalType 動物類型
     * @return 動物實例
     * @throws IllegalArgumentException 當動物類型未知時拋出
     */
    public static Animal createAnimal(String animalType) {
        switch (animalType.toLowerCase()) {
            case "cat":
                return new Cat();
            case "dog":
                return new Dog();
            case "bird":
                return new Bird();
            default:
                throw new IllegalArgumentException("未知的動物類型: " + animalType);
        }
    }
} 

File: Creational/SimpleFactory/cpp/AnimalFactory.cpp
#include "AnimalFactory.h"
#include "Cat.h"
#include "Dog.h"
#include "Bird.h"
#include <stdexcept>

std::unique_ptr<Animal> AnimalFactory::createAnimal(const std::string& animalType) {
    if (animalType == "cat") {
        return std::make_unique<Cat>();
    } else if (animalType == "dog") {
        return std::make_unique<Dog>();
    } else if (animalType == "bird") {
        return std::make_unique<Bird>();
    } else {
        throw std::invalid_argument("未知的動物類型: " + animalType);
    }
} 